packages/algorand-js/src/interpreter/opcode-list.ts

```diff
@@ -530,8 +549,22 @@ export class Concat extends Op {
 // push the substring result. If N < M, or either is larger than the string length,
 // the program fails
 export class Substring extends Op {
+  readonly start: bigint;
+  readonly end: bigint;
+
+  constructor (start: bigint, end: bigint) {
+    super();
+    this.start = start;
+    this.end = end;
+  }
```

Comment:
what is start and end in this case ? , In `substring3` we get the input from user

Reply:
Check opcode specification: [https://developer.algorand.org/docs/reference/teal/opcodes/?query=mulw#substring]()
`substring3` takes start and end from stack. But with `substring` you provide additional two params (start, end)
Eg
```
...
byte "Algorand"
substring 0 4       // pushes "Algo" to stack
...
```

Reply:
yes